saleor/graphql/order/mutations/fulfillments.py

```diff
@@ -622,22 +645,124 @@ class Meta:
         error_type_class = OrderError
         error_type_field = "order_errors"
 
+    @classmethod
+    def _check_shipping_costs(cls, payments_to_refund):
+        num_of_occurances = 0
+        for payment in payments_to_refund:
+            if payment.get("include_shipping_costs"):
+                num_of_occurances += 1
+
+        if num_of_occurances > 1:
+            raise ValidationError(
+                {
+                    "include_shipping_costs": ValidationError(
+                        "Shipping costs cannot be included in more than one payment.",
+                        code=OrderErrorCode.DUPLICATED_INPUT_ITEM,
+                    )
+                }
+            )
+
+    @classmethod
+    def _check_payments_belong_to_order(cls, order, payments):
+        improper_payments_ids = []
+        order_payments_ids = order.payments.values_list("id", flat=True)
+        for item in payments:
+            if not item["payment"].id in order_payments_ids:
+                improper_payments_ids.append(item["payment"].id)
+
+        if improper_payments_ids:
+            improper_payments_global_ids = [
+                graphene.Node.to_global_id("Payment", payment_id)
+                for payment_id in improper_payments_ids
+            ]
+            raise ValidationError(
+                {
+                    "payments_to_refund": ValidationError(
+                        "These payments do not belong to the order.",
+                        code=OrderErrorCode.PAYEMENTS_DO_NOT_BELONG_TO_ORDER,
+                        params={"payments": improper_payments_global_ids},
+                    )
+                }
+            )
+
+    @classmethod
+    def _check_order_has_single_payment(cls, order):
+        if order.payments.count() > 1:
+            raise ValidationError(
+                {
+                    "amount_to_refund": ValidationError(
+                        "It is not possible to use the amount field "
+                        "for orders with multiple payments.",
+                        code=OrderErrorCode.ORDER_HAS_MULTIPLE_PAYMENTS,
+                    )
+                }
+            )
+
+    @classmethod
+    def _get_payments_to_refund(
+        cls, info, order, payments_to_refund, amount_to_refund, include_shipping_costs
+    ):
+        if payments_to_refund:
+            cls._check_shipping_costs(payments_to_refund)
+            payments = []
+            for item in payments_to_refund:
+                payment = cls.get_node_or_error(info, item["payment_id"])
```

Comment:
Can we fetch payments in bulk? `.objects.in_bulk(keys)`

Reply:
Done